crepecake/src/main/java/com/google/cloud/tools/crepecake/cache/CacheChecker.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.cache;
+
+import com.google.cloud.tools.crepecake.image.ImageLayers;
+import com.google.cloud.tools.crepecake.image.ReferenceLayer;
+import java.io.File;
+import java.io.IOException;
+import java.util.Set;
+
+/** Checks if cached data is outdated. */
+public class CacheChecker {
+
+  private final Cache cache;
+
+  public CacheChecker(Cache cache) {
+    this.cache = cache;
+  }
+
+  /** @return true if the base image layers exist in the cache; false otherwise */
+  public boolean areBaseImageLayersCached(ImageLayers<ReferenceLayer> baseImageLayers)
+      throws CacheMetadataCorruptedException {
+    // Gets the base image layers.
+    ImageLayers<CachedLayerWithMetadata> cachedLayers =
+        cache.getMetadata().filterLayers().byType(CachedLayerType.BASE).filter();
+
+    if (cachedLayers.size() < baseImageLayers.size()) {
+      return false;
+    }
+
+    for (ReferenceLayer baseImageLayer : baseImageLayers) {
+      if (!cachedLayers.has(baseImageLayer.getBlobDescriptor().getDigest())) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Checks all cached layers built from the source files to see if the source files have been
+   * modified since the newest layer build.
+   *
+   * @param sourceFiles the source files to check
+   * @return true if no cached layer exists that are up-to-date with the source files; false
+   *     otherwise.
+   */
+  public boolean areSourceFilesModified(Set<File> sourceFiles)
```

Comment:
That's pretty expensive huh...

Reply:
Yea we could check based on checksum, but that would involve reading the full file. Last modified time may not be completely reliable since other things can modify the last modified time, but that might be the tradeoff - we make the assumption that the files in our cache are only touched by us.

Reply:
And how expensive is reading the full file in the environment where this will run. Can we estimate this in seconds, milliseconds, etc.?

Reply:
Thanks. The next question is will this be run in an environment where the difference between 0.5s and 14s is relevant. If it's a user facing job, yes. If a batch job probably not.